### PR TITLE
Project from

### LINKED ISSUE #22: Project Form

### PR DESCRIPTION
**Pull request by robsel118**
resolves #22 and  #39 

- Added creation and update form in `ProjectForm` view with `ProjectItem` component 
- Added API helper functions
- Added creation and update of updates using `UpdateForm` Component
- Added data fetching to API, no more dummy data
- Corrected project routes endpoint
- removed isAdmin from skill route post Function
- Fixed few design issues
- Replace LinkedIn Button
- Added `SearchView` page containing `SearchButton`
- Added Gravatar to project view
- Implemented deletion of projects and updates
- Fixed project delete API to only be deleted by owner

### CODE REVIEW

**Comment by MateyByrd** [comment]
Tried out the project creation, worked fine! When posting an update immediately after the backend crashed and I am not quite sure why. It gave me this error

```
events.js:167
      throw er; // Unhandled 'error' event
      ^

RangeError: Maximum call stack size exceeded
    at RegExp.test (<anonymous>)
    at isKey (/.../andromeda/server/node_modules/lodash.get/index.js:561:24)
```

**Comment by robsel118** [comment]
> 
> 
> @robsel118 I Tried out the project creation, worked fine! When posting an update immediately after the backend crashed and I am not quite sure why. It gave me this error
> 
> ```
> events.js:167
>       throw er; // Unhandled 'error' event
>       ^
> 
> RangeError: Maximum call stack size exceeded
>     at RegExp.test (<anonymous>)
>     at isKey (/.../andromeda/server/node_modules/lodash.get/index.js:561:24)
> ```
> 
> I also noted that the `POST /api/v1/projects/...id.../updates` request had no parameters in its body. After restarting the backend all seems to work correctly.
> 
> **Edit** The problems seems to be reproducible by creating a project (with all fields filled in) and then without reloading the page trying to post an update.

hhmm, I will look into fixing this

File: application/src/views/Project.vue [hunk 1]
```diff
@@ -1,116 +1,183 @@
...
     }
   },
   computed: {
-    getProjectData () {
-      // $route.param.id
-      return ''
+    checkMembership () {
+      if (getUser() !== null &&
+      (this.project.members.includes(getUser()._id) ||
+      this.project.owner._id === getUser()._id)) {
```
Comment (MateyByrd) [inline application/src/views/Project.vue]: Here I would also suggest changing the order of the checks as the first one is more expensive than the last one.

File: application/src/views/Project.vue [hunk 2]
```diff
@@ -1,116 +1,183 @@
...
+        return true
```
Comment (MateyByrd) [inline application/src/views/Project.vue]: As you are returning `true` in the `true` case of the `if`-statement and `false` in the `false` case. I would suggest returning the actual check in the `if`-statement.

File: application/src/views/Project.vue [hunk 3]
```diff
@@ -1,116 +1,183 @@
...
+      } else {
+        return false
+      }
+    },
+    checkFollowed () {
+      if (getUser() != null) {
```
Comment (MateyByrd) [inline application/src/views/Project.vue]: I think this should have another `=`
```suggestion
      if (getUser() !== null) {
```

File: application/src/views/ProjectForm.vue [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  name: 'ProjectForm',
+  components: { ProjectItem },
+  data () {
+    return {
+      project: {},
+      display: false
+    }
+  },
+  async mounted () {
+    if (getUser() == null) {
```
Comment (MateyByrd) [inline application/src/views/ProjectForm.vue]:
```suggestion
    if (getUser() === null) {
```

File: application/src/views/ProjectForm.vue [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      window.location = '/project/' + this.$route.params.id
```
Comment (MateyByrd) [inline application/src/views/ProjectForm.vue]:
```suggestion
      window.location = `/project/${this.$route.params.id}`
```

File: application/src/views/Projects.vue [hunk 6]
```diff
@@ -1,35 +1,74 @@
...
     SideModal
+  },
+  async mounted () {
+    this.projects = await getProjects()
+  },
+  methods: {
+    async createNewProject () {
+      this.newProject.owner = getUser()
+      var project = await createProject(this.newProject)
+      window.location = '/project/' + project._id + '/edit'
```
Comment (MateyByrd) [inline application/src/views/Projects.vue]:
```suggestion
      window.location = `/project/${project._id}/edit`
```

File: server/routes/api/v1/projects.js [hunk 7]
```diff
@@ -75,10 +76,10 @@ router.put('/:id',
...
     let id = req.params.id;
 
-    delete req.body.createdOn;
-    delete req.body.members;
-    delete req.body.requestedSkills;
-    delete req.body.updates;
+    // delete req.body.createdOn;
+    // delete req.body.members;
+    // delete req.body.requestedSkills;
+    // delete req.body.updates;
```
Comment (MateyByrd) [inline server/routes/api/v1/projects.js]: I would suggest removing the lines if they are no longer required

File: application/src/views/Project.vue [hunk 8]
```diff
@@ -1,116 +1,183 @@
...
+        <button class="btn btn-primary join" v-if="!checkMembership"> Join the Adventure</button>
+        <button @click="follow()" class="btn follow" v-if="checkFollowed" title="Add it to your timeline"><font-awesome-icon class="blue-title" icon="thumbtack" /> Follow!</button>
       </div>
     </div>
     <div class="grid-item section__6">
       <hr>
       <p
-        :class="{ active: descriptionVisible}"
-        @click="toggleDesciptionvisibility(true)"
+        :class="{ 'active': subPage == 0 }"
```
Comment (MateyByrd) [inline application/src/views/Project.vue]:
```suggestion
        :class="{ 'active': subPage === 0 }"
```

File: application/src/views/Project.vue [hunk 9]
```diff
@@ -1,116 +1,183 @@
...
+        @click="toggleDesciptionvisibility(0)"
         class="toggle">Description</p>
       <p
-        :class="{ active: !descriptionVisible}"
-        @click="toggleDesciptionvisibility(false)"
+        :class="{ 'active': subPage == 1 }"
```
Comment (MateyByrd) [inline application/src/views/Project.vue]:
```suggestion
        :class="{ 'active': subPage === 1 }"
```

File: application/src/views/Project.vue [hunk 10]
```diff
@@ -1,116 +1,183 @@
...
+        @click="toggleDesciptionvisibility(1)"
         class="toggle">Updates</p>
       <hr>
     </div>
-    <div v-show="descriptionVisible" class="grid-item project-descriptions section__6">
+    <div v-show="subPage == 0" class="grid-item project-descriptions section__6">
```
Comment (MateyByrd) [inline application/src/views/Project.vue]:
```suggestion
    <div v-show="subPage === 0" class="grid-item project-descriptions section__6">
```

File: application/src/components/LinkedInLoginButton.vue [hunk 11]
```diff
@@ -1,6 +1,9 @@
 <template>
   <div class="btn linked-in-login-button" @click="onPress">
-    <img src="../assets/img/linked-in/Sign-in-Large---Default.png">
+    <button class="btn btn-primary">
```
[the remark this answers is not in the document]
Comment (robsel118) [reply application/src/components/LinkedInLoginButton.vue]: removed

File: application/src/views/Project.vue (hunk 2 above)
Comment (robsel118) [reply application/src/views/Project.vue]: +1

File: application/src/views/Project.vue [hunk 12]
```diff
@@ -1,116 +1,183 @@
...
-          date: 'some date',
-          text: 'some text'
-        },
-        {
-          date: 'some date',
-          text: 'some text'
-        }
-        ]
-      }
+      project_id: this.$route.params.id,
```
[the remark this answers is not in the document]
Comment (robsel118) [reply application/src/views/Project.vue]: I have tried this variant, but it does not work for me. I have looked into your other views where you implemented this but nothing

**Pull request merged**